tools/c7n_azure/c7n_azure/session.py

```diff
@@ -169,19 +172,54 @@ def load_auth_file(self, path):
                 tenant=data['credentials']['tenant']
             ), data['subscription'])
 
-    def get_auth_string(self):
-        if type(self.credentials) is not ServicePrincipalCredentials:
+    def get_functions_auth_string(self):
+        """
+        Build auth json string for deploying
+        Azure Functions.  Look for dedicated
+        Functions environment variables or
+        fall back to normal Service Principal
+        variables.
+
+        """
+
+        function_auth_variables = [
+            'AZURE_FUNCTION_TENANT_ID',
+            'AZURE_FUNCTION_CLIENT_ID',
+            'AZURE_FUNCTION_CLIENT_SECRET'
+        ]
+
+        # Check for override on subscription ID
+        if 'AZURE_FUNCTION_SUBSCRIPTION_ID' in os.environ:
+            function_subscription_id = os.environ['AZURE_FUNCTION_SUBSCRIPTION_ID']
+        else:
+            function_subscription_id = self.subscription_id
+
+        # Use dedicated function env vars if available
+        if all(k in os.environ for k in function_auth_variables):
+            auth = {
+                'credentials':
+                    {
+                        'client_id': os.environ['AZURE_FUNCTION_CLIENT_ID'],
+                        'secret': os.environ['AZURE_FUNCTION_CLIENT_SECRET'],
+                        'tenant': os.environ['AZURE_FUNCTION_TENANT_ID']
+                    },
+                'subscription': function_subscription_id
+            }
+
+        elif type(self.credentials) is ServicePrincipalCredentials:
+            auth = {
+                'credentials':
+                    {
+                        'client_id': os.environ['AZURE_CLIENT_ID'],
+                        'secret': os.environ['AZURE_CLIENT_SECRET'],
```

Comment:
nit: should we move AZURE_CLIENT_SECRET and other strings to constants?

Reply:
It was my instinct to move every string in this file (~10 of them) to constants.py but I went and looked at the rest of the codebase and that is not how it is done.  In similar functionality in CC core all the environment variables are inline and used over and over again.

Reply:
a constants file would be reasonable and possibly more discoverable, we could refactor on core, probably worth filing an issue separately for that.